encoding.bs

```diff
@@ -1305,6 +1349,399 @@ must run these steps:
+
+  <pre class=example id=example-textdecoderstream-writable><code class=lang-javascript>
+var decoder = new TextDecoderStream(encoding);
+byteReadable
+  .pipeThrough(decoder)
+  .pipeTo(textWritable);</code></pre>
+
+  <p>If the <a for=TextDecoderCommon>error mode</a> is "<code>fatal</code>" and
+  <a for=TextDecoderCommon>encoding</a>'s <a for=/>decoder</a> returns <a>error</a>, both
+  {{GenericTransformStream/readable}} and {{GenericTransformStream/writable}} will be errored with a
+  {{TypeError}}.
+</dl>
+
+<p>The
+<dfn constructor for=TextDecoderStream id=dom-textdecoderstream><code>TextDecoderStream(<var>label</var>,
+<var>options</var>)</code></dfn> constructor, when invoked, must run these steps:
+
+<ol>
+ <li><p>Let <var>encoding</var> be the result of <a>getting an encoding</a> from <var>label</var>.
+
+ <li><p>If <var>encoding</var> is failure or <a>replacement</a>, then <a>throw</a> a {{RangeError}}.
+
+ <li><p>Let <var>dec</var> be a new {{TextDecoderStream}} object.
+
+ <li><p>Set <var>dec</var>'s <a for=TextDecoderCommon>encoding</a> to <var>encoding</var>.
+
+ <li><p>If <var>options</var>'s <code>fatal</code> member is true, then set <var>dec</var>'s
+ <a for=TextDecoderCommon>error mode</a> to "<code>fatal</code>".
+
+ <li><p>If <var>options</var>'s <code>ignoreBOM</code> member is true, then set <var>dec</var>'s
+ <a for=TextDecoderCommon>ignore BOM flag</a>.
+
+ <li>
+  <p>Set <var>dec</var>'s <a for=TextDecoderStream>decoder</a> to a new <a for=/>decoder</a>
+  for <var>dec</var>'s <a for=TextDecoderCommon>encoding</a>, and set <var>dec</var>'s
+  <a for=TextDecoderStream>stream</a> to a new <a for=/>stream</a>.
+
+ <li><p>Let <var>startAlgorithm</var> be an algorithm that takes no arguments and returns nothing.
+
+ <li><p>Let <var>transformAlgorithm</var> be an algorithm which takes a <var>chunk</var> argument
+ and runs the <a>decode and enqueue a chunk</a> algorithm with <var>dec</var> and
+ <var>chunk</var>.
+
+ <li><p>Let <var>flushAlgorithm</var> be an algorithm which takes no arguments and runs the <a>flush
+ and enqueue</a> algorithm with <var>dec</var>.
+
+ <li><p>Let <var>transform</var> be the result of calling
+ <a abstract-op>CreateTransformStream</a>(<var>startAlgorithm</var>, <var>transformAlgorithm</var>,
+ <var>flushAlgorithm</var>).
+
+ <li><p>Set <var>dec</var>'s <a for=GenericTransformStream>transform</a> to <var>transform</var>.
+
+ <li><p>Return <var>dec</var>.
+</ol>
+
+<p>The <dfn id=concept-tds-decode-and-enqueue>decode and enqueue a chunk</dfn> algorithm, given a
+{{TextDecoderStream}} <var>dec</var> and a <var>chunk</var>, runs these steps:
+
+<ol>
+ <li><p>Let <var>bufferSource</var> be the result of
+ <a lt="converted to an IDL value">converting</a> <var>chunk</var> to a {{BufferSource}}. If this
+ throws an exception, then return a promise rejected with that exception.
+
+ <li><p><a>Push</a> a <a lt="get a copy of the buffer source">copy of</a> <var>bufferSource</var> to
```

Comment:
In 6a65ad5da12cd43b24ee9459ea793d285f424912, we explicitly checked for `IsDetachedBuffer(arrayBuffer)` and returned a rejected promise if the buffer was detached. However, none of the three algorithms that make up [_convert to a BufferSource_](https://heycam.github.io/webidl/#es-buffer-source-types) check for detached buffers.

When [_get a copy of the buffer source_](https://heycam.github.io/webidl/#ref-for-dfn-get-buffer-source-copy%E2%91%A0) encounters a detached buffer, it throws a `TypeError` in step 7. This is not what we want: we want to return a rejected promise instead!

I think the easiest fix would be to add the following line to this step, similar to the previous step:
> If this throws an exception, then return a promise rejected with that exception.

Reply:
Great catch, thanks.

I did what you proposed. This has the added benefit that we are always in sync with the checks and behaviour provided by WebIDL.